osutil/mkdir.go

```diff
@@ -0,0 +1,164 @@
+	// If false (the default), return an error if the parent directory is missing.
+	// If true, any missing parent directories of the given path are created
+	// (with the same permissions and owner/group as the leaf directory).
+	MakeParents bool
+
+	// If false (the default), return an error if the target directory already exists.
+	// If true, don't return an error if the target directory already exists (unless
+	// it's not a directory).
+	ExistOK bool
+
+	// If false (the default), no explicit chmod is performed. In this case, the permission
+	// of the created directories will be affected by umask settings.
+	//
+	// If true, perform an explicit chmod on any directories created.
+	Chmod bool
+
+	// If false (the default), no explicit chown is performed.
+	// If true, perform an explicit chown on any directories created, using the UserID
+	// and GroupID provided.
+	Chown   bool
+	UserID  sys.UserID
+	GroupID sys.GroupID
+}
+
+// Mkdir creates a directory at path with the permissions perm and the options provided.
+// If options is nil, it is treated as &MkdirOptions{}.
+func Mkdir(path string, perm os.FileMode, options *MkdirOptions) error {
+	if options == nil {
+		options = &MkdirOptions{}
+	}
+	mu.Lock()
+	defer mu.Unlock()
+
+	path = filepath.Clean(path)
+
+	if s, err := os.Stat(path); err == nil {
+		// If path exists but not as a directory, return a "not a directory" error.
+		if !s.IsDir() {
+			return &os.PathError{
+				Op:   "mkdir",
+				Path: path,
+				Err:  syscall.ENOTDIR,
+			}
+		}
+
+		// If path exists as a directory, and ExistOK option is set, do nothing.
+		if options.ExistOK {
+			return nil
+		}
+
+		// If path exists but ExistOK option isn't set, return a "file exists" error.
+		return &os.PathError{
+			Op:   "mkdir",
+			Path: path,
+			Err:  syscall.EEXIST,
+		}
+	}
+
+	// If path doesn't exist, create it.
+	return mkdirAll(path, perm, options)
+}
+
+// create directories recursively
+func mkdirAll(path string, perm os.FileMode, options *MkdirOptions) error {
+	// if path exists
+	if s, err := os.Stat(path); err == nil {
+		if s.IsDir() {
+			return nil
+		}
+
+		// If path exists but not as a directory, return a "not a directory" error.
+		return &os.PathError{
+			Op:   "mkdir",
+			Path: path,
+			Err:  syscall.ENOTDIR,
+		}
+	}
+
+	// If path doesn't exist, and MakeParents is specified in options,
+	// create all directories recursively.
+	if options.MakeParents {
+		parent := filepath.Dir(path)
+		if parent != "/" {
+			if err := mkdirAll(parent, perm, options); err != nil {
+				return err
+			}
+		}
+	}
+
+	// If path doesn't exist, and MakeParents isn't specified in options,
+	// create a single directory.
+	return mkdir(path, perm, options)
+}
+
+// Create a single directory and perform chown/chmod operations according to options.
+func mkdir(path string, perm os.FileMode, options *MkdirOptions) error {
+	cand := path + ".mkdir-new"
+
+	if err := os.Mkdir(cand, perm); err != nil && !os.IsExist(err) {
```

Comment:
beyond this point the candidate is left behind on errors, isn't it?

Reply:
Hi, I did not quite understand your question, what do you mean by "the candidate is left behind on errors"? Could you elaborate? (FYI, this logic too is from [the existing code](https://github.com/canonical/snapd/blob/master/osutil/mkdirallchown.go#L35).)

Reply:
What I mean is that a temporary directory `<name>.mkdir-new` gets created here, but there is no defererred cleanup on errors, so should any error occur later in Chmod/Chown/Rename, that directory will be left behind. 

And yes, it' looks like the problem exists in the code we have in master, so even more reason to fix when introducing a new API.